### ISSUE TITLE
Different indentation level changes behaviour of the code

### ISSUE THREAD

**Issue by smoothdeveloper** [dotnet/fsharp#1019]
``` fsharp
let foo t = "(" + t + ")"

let tttt =
  let text =
    match Some "" with
    | Some t ->
      match t with
      | "" ->
        "a"
      | "b" ->
        "b"
      |> foo
    | _ -> "fooo"
  text  

let tttt' =
    let text =
        match Some "" with
        | Some t ->
            match t with
            | "" ->
                "a"
            | "b" ->
                "b"
            |> foo
        | _ -> "fooo"
    text
```

In `foo` I use two spaces for indentation, in `foo'` four spaces are used.

The resulting expressions are different while I'd had expected those to be the same.

I tried to find something which would explain that in the language specification (which I haven't thoroughly read) but failed, and the behaviour seems to fail me at "principle of least surprise" so submitting this as an issue to get clarification.

Thanks.

**Comment by radekm** [comment]
IMO it seems correct - see the specification [15.1.9 Exceptions to the Offside Rules](http://fsharp.org/specs/language-spec/4.0/FSharpSpec-4.0-latest.pdf#285) - the first exception:

> An infix token may be offside by the size of the token plus one.

Here is a similar example:

```fsharp
let (||>) = (|>)

let () =
    let x =
        match "" with
        | "" -> 1
        | _ ->
            4
        |> (+) 9
    printfn "%A" x
```

The code prints `10` because the exception doesn't apply (the expression in the second branch of match is `4`).

But if you replace `|>` by `||>` or decrease the indentation of the line with `4` by one space the result will change to `1` because the exception to the offside rule will apply (the expression in the second branch of match will be `4 |> (+) 9`).

**Comment by smoothdeveloper** [comment]
@radekm, thanks for bringing the relevant spec bits.

This seems like a serious gotcha to (try to) keep in mind while writing F# code.

**Comment by dsyme** [comment]
@smoothdeveloper  Please consider opening a suggestion on http://fslang.uservoice.com to request a warning for the first of these cases.  It would be great if you could formulate an exact proposed rule (e.g. one that would cover the interaction with "match", but no more, unless you identify other similar interactions).

**Comment by smoothdeveloper** [comment]
@dsyme you are right, a warning might make the most sense, but I'm trying to wrap my head around all the cases shown in the spec.

Wouldn't it be simpler/better to have a `strict-indent` compiler flag which disable the subset of those rules related to expressions (don't want to impact the type related and `for` constructs).

**Comment by dsyme** [comment]
@smoothdeveloper We try really, really hard not to add compiler flags that "fork" the language (or even restrict it).  It's possible that adding the warning to FSharpLint or Fantomas or FSharp.Compiler.Service may be the right place.

**Comment by smoothdeveloper** [comment]
Reopening in case this can be a warning: we should always recommend putting the thing to be piped between parens, this still behaves the same as suggestion: https://github.com/fsharp/fslang-suggestions/issues/806.

**Comment by brianrourkeboll** [comment]
@smoothdeveloper It would be relatively straightforward—at least as an analyzer, if not during type checking—to emit a diagnostic when a `SynMatchClause` ended in a `SynExpr.App` of an infix operator (or specifically a pipe operator), the operator's right-hand argument was not on the same line as the left-hand argument, and the start range of the operator was leftward of the left-hand operand (or was on the same column as the match clause's `|`, as in your example).
